File: gobang/gobang_python/AI/ai.py
import logging

from .static import *
import AI.config as config
from .negamax import deepenAll
from .board import Board

_aiLogger = logging.getLogger("GobangAI")


class WinError(Exception):
    pass


class AI:
    def __init__(self):
        _aiLogger.warning("init AI")
        self.board = Board()

    def begin(self):
        _aiLogger.info("AI's turn")
        if len(self.board.allSteps) > 1:
            # p = opening(self.board)
            ...
        p = deepenAll(self.board, None, config.searchDepth)
        _aiLogger.debug(f"AI's next step is Place({p.x}, {p.y}))")
        self.board.put(p, Role.COMPUTER, log=True)

        win = self.board.isWin(log=True)
        if win:
            role, p = win
            _aiLogger.info(f"{role.name()} at Place(x={p[0]}, y={p[1]})")
            raise WinError(role)

        return p

    def turn(self, place):
        _aiLogger.info(f"human's next step is Place({place.x}, {place.y})")
        self.set(place, Role.HUMAN)

        win = self.board.isWin(log=True)
        if win:
            role, p = win
            _aiLogger.info(f"{role.name()} at Place(x={p[0]}, y={p[1]})")
            raise WinError(role)

        return self.begin()

    def set(self, place, role):
        _aiLogger.debug(f"set Place({place.x}, {place.y}) to {role.name()}")
        self.board.put(place, role, log=True)

File: gobang/gobang_c++/AI/ai.cpp
//
// Created by ycm on 2023/12/9.
//

#include "ai.h"

Logger aiLogger("GobangAI");

//Computer::Computer():board() {}

Place Computer::setComputer() {
    aiLogger.info("Computer is thinking...");
    if(this->board.allSteps.size()>1) {
        //TODO: 随机开局
    }
    auto p = deepenAll(this->board, Role::COMPUTER, config::searchDepth);
    this->board.put(p, Role::COMPUTER, true);

    this->isWin(true);
    return p;
}

Place Computer::setHuman(Place &place) {
    aiLogger.info(std::string("human's next step is "+place.toString()));

    if(!(this->set(place, Role::HUMAN)) || this->isWin(true)!=Role::NONE){
        return *(new Place{-1, -1});
    }else{
        return this->setComputer();
    }
}

bool Computer::set(Place &place, Role role) {
    return this->board.put(place, role, true);
}

Role Computer::isWin(bool log/*=false*/) {
    auto Win = this->board.isWin(log);
    if(Win!=Role::NONE){
        aiLogger.info((Win==Role::COMPUTER)?"Computer wins!":"Human wins!");
        // TODO: Game over
    }
    this->win = Win;
    return Win;
}
